### DISCUSSION TITLE
discussion: bidirectionality

### DISCUSSION THREAD

**Discussion by justin2004** [SPARQL-Anything/sparql.anything#390] [Ideas]
we can consume from, say, a csv file but it might be interesting to think about inserting into a csv file too.

e.g.

/mnt/test3.csv:
```
name,friend,age
"fred","joe",10
"sally","jim",
"will","bill",""
"joe","N/A",""
```

and then something like this:
```
curl --silent 'http://localhost:3000/sparql.anything'  \
--header "Accept: text/csv" \
--data-urlencode 'query=
PREFIX xyz: <http://sparql.xyz/facade-x/data/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX fx: <http://sparql.xyz/facade-x/ns/>
delete { ?s xyz:name "will" }
insert { ?s xyz:name "willomena" }
where {
 SERVICE <x-sparql-anything:> {
   fx:properties fx:location "/mnt/test3.csv" ;
                 fx:csv.headers "true" ;
.
   ?s xyz:name "will" .
 }
}'
```

could turn /mnt/test3.csv into:
```
"fred","joe",10
"sally","jim",
"willomena","bill",""
"joe","N/A",""
```

**Comment by justin2004** [comment]
i've since learned this is called "lifting and lowering."

**Comment by luigi-asprino** [comment]
Hi,

I just stepped into [SPARQL Transformer](https://github.com/D2KLab/sparql-transformer). 
Essentially, the tool allows to define prototypes for JSON objects. 
The prototypes are used for generating a SPARQL queries which serve in turn to instantiate the prototype.
I think that a similar approach can be implemented for "lowering" data.

**Comment by enridaga** [comment]
Nice! The SPARQL Transformer approach could be used to write back to JSON, then. Of course, we don't want to ask users to write those 'lowering' mappings but those should be produced automatically from the facade+format combination. Is there a chance to abstract this approach to any format, defining prototypes automatically for XML, HTML, etc ... so that we can support SPARQL Update on those potentially any format? I mean, the prototyping part is really about projecting the triple in the original format... 

wow, we are really flying high here!

**Comment by justin2004** [comment]
XSPARQL [specified](https://www.w3.org/Submission/xsparql-language-specification/#sec:liftlow) this functionality.
[This](https://github.com/semantalytics/xsparql) looks like it was one of the implementations.

I wonder if the project lost steam because it only supported XML?

**Comment by luigi-asprino** [comment]
This is an interesting discussion we need to resume, but, maybe, better to move it to the discussion section and then open specific issues once we start implementing this feature.

**Comment by VladimirAlexiev** [comment]
https://ci.mines-stetienne.fr/sparql-generate/ (https://github.com/sparql-generate/) is another tool that's very powerful for generating all kinds of text from RDF or SPARQL.
